lib/online_order.rb

```diff
@@ -0,0 +1,109 @@
+require_relative 'Order'
+require 'csv'
+
+#The fact that I am redefining the methods OnlineOrder inherits from Order is why I am getting all the warning messages when I run rake
+
+module Grocery
+  class OnlineOrder < Order
+      attr_reader :customer_id, :status, :customer
+
+    def initialize(id, products, customer_id, status = :pending)
+      super(id, products)
+      #took out customer_id since I am using @customer now
+      # @customer_id = customer_id
+      @status = status
+      @customer = Grocery::Customer.find(customer_id) #so that we use an instance of the Customer class instead of just the customer_id in the csv file to access data about each customer
+    end #initialize
+
+#Modifying the total method so that a shipping cost will only be added if super returns a value grater than 0 (meaning that there are products in the order)
+    def total
+      if super == 0
+        return 0
+      else
+        return super + 10
+      end #if/else
+    end #total
+
+#You cannot add a product to an online order if @status of the order is :processing, :shipped, or :completed
+#other then adding the functionality above this method just calls the super method from the Order class
+    def add_product(products, price)
+      if @status == :processing || @status == :shipped || @status == :completed
+        raise ArgumentError.new("Error: you cannot add to you order because it is allready #{@status}")
+      else
+        super
+      end #if/else
+    end #add_product
+
+    #TODO is it OK that I just totally rewrote the self.all method? Should I have used super? I didn't know how to do this cause they changes I would have needed to make to the file would be within the functionality of super....
+    #Similar to self.all in Order, but we also have the customer_id and the status in the csv file.
```

Comment:
Good question. Inheritance starts to get tricky when you're working with class methods - many times the syntax for things like this won't quite match up.

One strategy you could try is to define two other class methods, `self.from_csv_line` and `self.file_name`.
- `self.file_name` just returns the name of the CSV file.
- `self.from_csv_line` takes a line of the CSV, parses it, and calls `new` with the correct arguments. It would behave a little differently for `Order` and `OnlineOrder`.

Then your definition of `self.all` would call those two methods.

In fact since loading from CSV is a _behavior_, using a module/mixin might be a better match here. That way you could include it in `Customer` as well.